packages/flutter/lib/src/widgets/scroll_notification.dart

```diff
@@ -0,0 +1,187 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/gestures.dart';
+import 'package:flutter/rendering.dart';
+
+import 'framework.dart';
+import 'basic.dart';
+import 'notification_listener.dart';
+import 'scrollable.dart' show Scrollable2, Scrollable2State;
+
+class ScrollableMetrics {
+  const ScrollableMetrics({
+    this.extentBefore,
+    this.extentInside,
+    this.extentAfter,
+  });
+
+  final double extentBefore;
+  final double extentInside;
+  final double extentAfter;
+
+  @override
+  String toString() {
+    return '$runtimeType(${extentBefore.toStringAsFixed(1)}..[${extentInside.toStringAsFixed(1)}]..${extentAfter.toStringAsFixed(1)}})';
+  }
+}
+
+abstract class ScrollNotification2 extends LayoutChangedNotification {
+  /// Creates a notification about scrolling.
+  ScrollNotification2({
+    Scrollable2State scrollable,
+  }) : axisDirection = scrollable.config.axisDirection,
+       metrics = scrollable.position.getMetrics(),
+       context = scrollable.context;
+
+  /// The direction that positive scroll offsets indicate.
+  final AxisDirection axisDirection;
+
+  Axis get axis {
```

Comment:
We have this function in a bunch of places.  Maybe factor out into a static function and have all these call through to it?